mla/threeml/sob_terms.py

```diff
@@ -0,0 +1,225 @@
+
+import dataclasses
+
+import numpy as np
+from scipy.interpolate import UnivariateSpline as Spline
+
+from .. import sob_terms
+from .. import sources
+from .. import params as par
+from . import spectral
+from . import data_handlers
+
+
+@dataclasses.dataclass
+class ThreeMLPSEnergyTerm(sob_terms.SoBTerm):
+    """Docstring"""
+
+    _energysobhist: np.ndarray
+    _sin_dec_idx: np.ndarray
+    _log_energy_idx: np.ndarray
+
+    def update_sob_hist(
+        self,
+        factory: sob_terms.SoBTermFactory
+    ) -> None:
+        """Docstring"""
+        self._energysobhist = factory.cal_sob_map()
+
+    @property
+    def params(self) -> par.Params:
+        """Docstring"""
+        return self._params
+
+    @params.setter
+    def params(self, params: par.Params) -> None:
+        """Docstring"""
+        self._params = params
+
+    @property
+    def sob(self) -> np.ndarray:
+        """Docstring"""
+        return self._energysobhist[self._sin_dec_idx, self._log_energy_idx]
+
+
+@dataclasses.dataclass
+class ThreeMLPSEnergyTermFactory(sob_terms.SoBTermFactory):
+    """Docstring"""
+
+    data_handler: data_handlers.ThreeMLDataHandler
+    source: sources.PointSource
+    spectrum: spectral.BaseSpectrum
+    _source: sources.PointSource = dataclasses.field(
+        init=False, repr=False)
+    _spectrum: spectral.BaseSpectrum = dataclasses.field(
+        init=False, repr=False, default=spectral.PowerLaw(1e3, 1e-14, -2))
+    _bg_sob: np.ndarray = dataclasses.field(init=False, repr=False)
+    _sin_dec_bins: np.ndarray = dataclasses.field(init=False, repr=False)
+    _log_energy_bins: np.ndarray = dataclasses.field(init=False, repr=False)
+    _bins: np.ndarray = dataclasses.field(init=False, repr=False)
+
+    def __post_init__(self) -> None:
+        """Docstring"""
+        self._sin_dec_bins = np.linspace(
+            -1, 1, 1 + self.config["sin_dec_bins"]
+        )
+        self._log_energy_bins = np.linspace(
+            *self.config["log_energy_bounds"],
+            1 + self.config["log_energy_bins"]
+        )
+        self.data_handler.reduced_reco_sim = (
+            self.data_handler.cut_reconstructed_sim(
+                self.source.location[1],
+                self.data_handler.config["reco_sampling_width"],
+            )
+        )
+        self._bins = np.array([self._sin_dec_bins, self._log_energy_bins])
+        self._init_bg_sob_map()
+
+    def __call__(
+        self, params: par.Params, events: np.ndarray
+    ) -> sob_terms.SoBTerm:
+        """Docstring"""
+        sin_dec_idx = np.searchsorted(
+            self._sin_dec_bins[:-1], events["sindec"]
+        )
+        log_energy_idx = np.searchsorted(
+            self._log_energy_bins[:-1], events["logE"]
+        )
+
+        # Get the bin that each event belongs to
+        try:
+            sin_dec_idx = (
+                np.searchsorted(self._sin_dec_bins[:-1], events["sindec"]) - 1
+            )
+        except ValueError:

```

Comment:
does this need to be here if we guarantee sindec exists?

Reply:
ok